[B] Add API for snow layer blocks. Addresses BUKKIT-5421

MisterVector opened the pull request:
Currently there is no way to get/set the various heights of a snow layer
block. This commit allows plugin authors to get and set the height of 
snow layer blocks.
### The Issue:

Presently, there is no API for manipulating height for snow layer blocks. This PR adds relevant API for controlling/viewing the height of a snow layer block.
### Justification for this PR:

It would be nice if bukkit plugin developers had access to the ability to manipulate snow layer block height. Such use could have interesting effects when creating a custom generator involving snow layer blocks at differing heights.
### PR breakdown:

This PR adds a material data class designed for snow layer blocks along with an enum for describing snow layer height. These are used for modifying snow height by data value manipulation.
### Testing results and materials:

I tested this out by creating a plugin that grabs a block's MaterialData, check if it is an instance of snow layer, then casts to that if so. I accept input on a data value from 0-7 indicating the data value for the snow layer height. /snow [height data value] is the only command implemented. Using /snow without arguments shows the height of the snow layer, and providing input as /snow 0-7 sets the height of the snow layer block. /snow addlayer and /snow removelayer will add to, or remove from, the current layer of the snow layer block, respectively. If the value is not a number, or is outside the range of the available snow layer data values, an error is given.

Source: http://pastebin.com/wGZneQ1S
Binary: https://dl.dropboxusercontent.com/u/13099063/TestPlugin.jar
### JIRA ticket

https://bukkit.atlassian.net/browse/BUKKIT-5421

MisterVector committed:
Add API for snow layer blocks. Addresses BUKKIT-5421

Currently there is no way to get/set the various heights of a snow layer
block. This commit allows plugin authors to get and set the height of
snow layer blocks.

spaceemotion commented:
I definitely like this, but what about adding functions like `addLayer` and `removeLayer` as well? Just a thought ...

MisterVector commented:
Sounds interesting. I suppose I could do that.

MisterVector commented:
There appears to be an 8th layer, when snow layers are "full" blocks. I will figure that out and add the addLayer() / removeLayer() methods as well.

MisterVector committed:
- Added missing 8th snow layer block
- Added addLayer() / removeLayer() methods to SnowLayer metadata class
to add and remove snow layers

MisterVector commented:
Updated PR description with a new link to the current source code. The binary has been updated at the current link.

jessestricker commented on src/main/java/org/bukkit/SnowLayerHeight.java:
```diff
@@ -0,0 +1,78 @@
+    LAYER_SEVEN(0x6),
+
+    /*
+     * The eight snow layer.
```
You misspelled `eighth`. I think this should be fixed before pulling :wink:

MisterVector replied:
Thanks! I noticed that as well. I will see if anything else needs fixing before making another commit.

MisterVector committed:
- Corrected a few things

CelticMinstrel commented:
It doesn't seem appropriate to use an enum for this.

MisterVector committed:
- Don't use an enum and use raw data instead

MisterVector commented:
No longer using an enum. Source URL on pastebin has been updated.

turt2live commented:
A replacement ticket to the mistakenly categorized ticket has been created: [5700](https://bukkit.atlassian.net/browse/BUKKIT-5700)

turt2live commented on src/main/java/org/bukkit/material/SnowLayer.java:
```diff
@@ -0,0 +1,108 @@
+     * @param height the height of the new snow layer
+     */
+    public void setHeight(int height) {
+        if (height < 8) {
```
This should be a `Validate` call instead. Also why does this accept an `int` and the getter returns a `byte`? Shouldn't they be consistent?

turt2live commented:
I'm not sure if a 0-based height is ideal for this situation. Generally 0 would imply "no height" however the snow layer is actually at 1/8th height. Thoughts?

CelticMinstrel commented:
I agree.

MisterVector commented:
I guess this will never be a thing.

kevinlul commented:
@MisterVector You should go check out the SpigotMC community for this.

mbax commented:
Or, the [SpongePowered](https://www.spongepowered.org/) community for an API approach that isn't as outdated as this PR. :)

MisterVector commented:
@cumandr I'd rather do this for Paper, not Spigot. Rep Paper!
